src/modules/wali/x86-64-linux/WaliImpl.v3

```diff
@@ -39,4 +47,63 @@ component WaliImpl {
 		fdmap.free(args);
 		return HostResult.Value1(Values.box_l(args));
 	}
+	def getPalignMmapAddr(memory: Memory) -> u64 {
+		var base_addr = memory.getMemBase64();
+		var punalign = base_addr + memory.num_pages * memory.num_bytes;
+		var palign = ((punalign + (X86_PAGE_SIZE-1)) * X86_PAGE_SIZE)/X86_PAGE_SIZE;
+		return palign;
+	}
+	def mmap(num: int, args: (Pointer, i32, i32, i32, i32, i64), fdmap: FileDescriptorMap, memory: Memory) -> HostResult {
+		//FIXME: casting madness
+		var pa_aligned_addr = getPalignMmapAddr(memory);
+		var mmap_addr = pa_aligned_addr + mmap_pagelen * X86_PAGE_SIZE;
+
+		var inc_wasm_pages = 0u64;
+		var num_pages = (u64.!(args.1) + X86_PAGE_SIZE - 1) / X86_PAGE_SIZE;
+		var extended_mmap_pagelen = mmap_pagelen + num_pages;
+
+		if (extended_mmap_pagelen > wasm_pagelen * WASM_TO_X86_PAGE) {
+			var new_wasm_pagelen = (extended_mmap_pagelen + WASM_TO_X86_PAGE - 1) / WASM_TO_X86_PAGE;
+			inc_wasm_pages = new_wasm_pagelen - wasm_pagelen;
+			// TODO: implement check if wasm memory needs to be expanded and it is safe
+		}
+		var new_args = (Pointer.NULL + long.view(mmap_addr), args.1, args.2, args.3 | LinuxConst.MAP_FIXED, args.4, args.5);
+		var sb = StringBuilder.new();
```

Comment:
Looks like some leftover tracing code here.

Also this, can be made a bit shorter with `Trace.OUT....ln()`.